### ISSUE TITLE
New Italian CNS lots ACe2021 ACx2021 and ST2021 unsupported

### ISSUE THREAD

**Issue by aquilarubra** [OpenSC/OpenSC#2313]
### Problem Description

New lots of Italian CNS, ACe2021 ACx2021 and ST2021, with expiry in 2027 are not supported.

**Comment by frankmorgner** [comment]
Please create a debug log of your problem as described in the [comments of our issue template](https://raw.githubusercontent.com/OpenSC/OpenSC/master/.github/ISSUE_TEMPLATE.md). You may also want to look at the hints on [how to write a good bug report](https://github.com/OpenSC/OpenSC/wiki/How-to-write-a-good-bug-report).

**Comment by aquilarubra** [comment]
I don't have those cards, so I cannot debug. But I work in the field and I see that people having those lots of cards report they are not working with opensc.

**Comment by frankmorgner** [comment]
Great, that you're in touch with those people. Please advice them how to provide basic debug information, see the links above.

**Comment by emmecii** [comment]
@frankmorgner I recently got one of those card (ST2021) i can report the problem

I got this error:

```
ivan@yang:~$ pkcs11-tool --test --login
Using slot 0 with a present token (0x0)
error: PKCS11 function C_Login failed: rv = CKR_GENERAL_ERROR (0x5)
Aborting.
```
I tried to login to INPS that require the CNS system ([Carta Nazionale dei Servizi)](https://en.wikipedia.org/wiki/Carta_nazionale_dei_servizi), it respond with "No valid certificate was provided for access: close the browser, check the prerequisites and try again with a new valid certificate. "

On Firefox, the card is read correctly trought a Bit4id MiniLector Evo reader

Note that the official driver for ST2021 is missing on the Ministry of Economy and Finance website (the supplier of the smartcard), you ca find the driver for Linux on the Sardinia region website [here](https://tscns.regione.sardegna.it/it/articoli/linux), but seems to be just a java applet to read the information inside the card

**Comment by fedetft** [comment]
Hello, I am adding to this report as I have a similar problem.
I am trying to get an Italian CNS to work, but I get the same error. Mine is not so new, but the error is the same.

`$ pkcs11-tool --test --login
Using slot 0 with a present token (0x0)
error: PKCS11 function C_Login failed: rv = CKR_GENERAL_ERROR (0x5)
Aborting.`

I ran the same command again with logging

`$ OPENSC_DEBUG=3 pkcs11-tool --test --login &> log-long.txt`

and can provide a log.

[log-long-sanitized.txt](https://github.com/OpenSC/OpenSC/files/6580969/log-long-sanitized.txt)

Also I am running OpenSC 0.20.3 on Ubuntu 20.04.2 LTS

**Comment by fedetft** [comment]
Tried building from source, (did a git clone of the master branch) just to see if the issue was already resolved, but get the same error.
Log is less verbose around the failure point, but it seems to be the same:

```
P:32496; T:0x140689282955328 01:25:51.056 [opensc-pkcs11] sec.c:200:sc_pin_cmd: called
P:32496; T:0x140689282955328 01:25:51.060 [opensc-pkcs11] sec.c:256:sc_pin_cmd: returning with: -1107 (Transmit failed)
```

**Comment by fedetft** [comment]
Mine is from ST, so it seems vendor-related

```
$ ./pkcs11-tool --list-slots
Available slots:
Slot 0 (0x0): Bit4id miniLector [miniLector-EVO] 00 00
  token label        : <omitted> (PIN CNS0)
  token manufacturer : IC: STMicroelectronics; mask:...
  token model        : PKCS#15 emulated
  token flags        : login required, PIN pad present, token initialized, PIN initialized, user PIN locked
  hardware version   : 0.0
  firmware version   : 0.0
  serial num         : <omitted>
  pin min/max        : 5/8
```

**Comment by frankmorgner** [comment]
Indeed, `Transmit failed` indicates a problem with the the reader or the card itself. Do you have a middleware from the vendor, which allows using the card on your platform?

**Comment by dengert** [comment]
OpenSC thinks the card reader has a PIN PAD: "PIN pad present: and  lines 2307 in your log is telling reader how to handle a pin verify command. 
```
PC/SC v2 pinpad block (32 bytes):
1E 1E 02 08 00 08 05 02 00 00 00 00 00 00 00 0D ................
00 00 00 00 20 00 10 08 FF FF FF FF FF FF FF FF .... ...........
```

But https://www.bit4id.com/en/smart-card-reader-minilector-evo/ shows a reader without a pinpad. 
https://resources.bit4id.com/files/documents/datasheet/miniLector%20EVO%20-%20DS%20-%20EN%20-%203.3.pdf
(It could be the vendor's reader driver is trying to emulate a pin pad internally. but there is no linux driver is installed for the reader.)
    
`man opensc.conf` shows how to not use the pin pad, `enable_pinpad = bool;` See in OpenSC source:
 `OpenSC/etc/opensc.conf.example.in` something like:

```
app default {
    reader_driver pcsc {
        enable_pinpad = false;
    }
}
```







Did your old card work with this reader?

Does you card work on windows with this reader? 

Can you try your card with another card reader?

Can you try some other card with this reader?

**Comment by fedetft** [comment]
Tried adding `enable_pinpad = false;` and now it is asking for my pin!
Unfortunately I don't have my pin, will be requesting it shortly, so I don't know if this fixes all the issues, but it definitely fixed the `C_Login failed` error, thanks!

Although I don't have pins, I could find an older CNS, and it is an ATMEL/Athena one. This older card behaves just like the new one, it gives `C_Login failed` unless `enable_pinpad = false;` is added to `opensc.conf`.

So it seems that, at least with my setup, it is not an error related to the card model, but it's related the the reader that incorrectly advertises it has a pinpad.

`dmesg` identifies it as:
```
[14427.585008] usb 1-2: new full-speed USB device number 8 using xhci_hcd
[14427.737841] usb 1-2: New USB device found, idVendor=25dd, idProduct=3111, bcdDevice= 1.11
[14427.737847] usb 1-2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[14427.737850] usb 1-2: Product: miniLector-EVO
[14427.737853] usb 1-2: Manufacturer: bit4id
```

And it has a suspicious serial number of 0. Maybe it's a knockoff of a real miniLector? I bought it on ebay...

**Comment by Andreabont** [comment]
> Tried adding `enable_pinpad = false;` and now it is asking for my pin!
> Unfortunately I don't have my pin, will be requesting it shortly, so I don't know if this fixes all the issues, but it definitely fixed the `C_Login failed` error, thanks!
> 
> Although I don't have pins, I could find an older CNS, and it is an ATMEL/Athena one. This older card behaves just like the new one, it gives `C_Login failed` unless `enable_pinpad = false;` is added to `opensc.conf`.
> 
> So it seems that, at least with my setup, it is not an error related to the card model, but it's related the the reader that incorrectly advertises it has a pinpad.
> 
> `dmesg` identifies it as:
> 
> ```
> [14427.585008] usb 1-2: new full-speed USB device number 8 using xhci_hcd
> [14427.737841] usb 1-2: New USB device found, idVendor=25dd, idProduct=3111, bcdDevice= 1.11
> [14427.737847] usb 1-2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
> [14427.737850] usb 1-2: Product: miniLector-EVO
> [14427.737853] usb 1-2: Manufacturer: bit4id
> ```
> 
> And it has a suspicious serial number of 0. Maybe it's a knockoff of a real miniLector? I bought it on ebay and it arrived yesterday...

I also have a miniLector EVO, and it works correctly. But i found some differences in dmesg:

```
[ 274.670312] usb 1-5: new full-speed USB device number 10 using xhci_hcd
[ 274.821858] usb 1-5: New USB device found, idVendor=25dd, idProduct=3111, bcdDevice= 1.00
[ 274.821859] usb 1-5: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[ 274.821860] usb 1-5: Product: miniLector EVO
[ 274.821861] usb 1-5: Manufacturer: BIT4ID
```

**Comment by aquilarubra** [comment]
Bit4id integrated ACe2021 and ACx2021. Apparently they don't use the standard format but the envelope is somewhat different.

Recently, Bit4id integrated ST2021 too (in this case, even the manufacturer's driver didn't work) for FVG region. I don't know the details though, aside that it needed some changes to the software.

**Comment by dengert** [comment]
Lines 13-27 in the log:
```
P:6268; T:0x140653054314560 00:22:35.279 [opensc-pkcs11] reader-pcsc.c:1500:pcsc_detect_readers: Bit4id miniLector [miniLector-EVO] 00 00:SCardConnect(SHARED): 0x00000000
P:6268; T:0x140653054314560 00:22:35.279 [opensc-pkcs11] reader-pcsc.c:1114:detect_reader_features: called
P:6268; T:0x140653054314560 00:22:35.279 [opensc-pkcs11] reader-pcsc.c:1116:detect_reader_features: Requesting reader features ... 
P:6268; T:0x140653054314560 00:22:35.279 [opensc-pkcs11] reader-pcsc.c:1137:detect_reader_features: Reader feature 06 found
P:6268; T:0x140653054314560 00:22:35.279 [opensc-pkcs11] reader-pcsc.c:1137:detect_reader_features: Reader feature 07 found
P:6268; T:0x140653054314560 00:22:35.280 [opensc-pkcs11] reader-pcsc.c:1137:detect_reader_features: Reader feature 0a found
P:6268; T:0x140653054314560 00:22:35.280 [opensc-pkcs11] reader-pcsc.c:1137:detect_reader_features: Reader feature 12 found
P:6268; T:0x140653054314560 00:22:35.280 [opensc-pkcs11] reader-pcsc.c:1165:detect_reader_features: Reader supports pinpad PIN verification
P:6268; T:0x140653054314560 00:22:35.280 [opensc-pkcs11] reader-pcsc.c:1175:detect_reader_features: Reader supports pinpad PIN modification
P:6268; T:0x140653054314560 00:22:35.280 [opensc-pkcs11] reader-pcsc.c:1093:part10_get_vendor_product: id_vendor=25dd id_product=3111
P:6268; T:0x140653054314560 00:22:35.280 [opensc-pkcs11] reader-pcsc.c:1215:detect_reader_features: Reader does not have a display.
P:6268; T:0x140653054314560 00:22:35.281 [opensc-pkcs11] reader-pcsc.c:1054:part10_detect_max_data: get dwMaxAPDUDataSize property returned 65536
P:6268; T:0x140653054314560 00:22:35.281 [opensc-pkcs11] reader-pcsc.c:1246:detect_reader_features: Reader supports transceiving 65536 bytes of data
P:6268; T:0x140653054314560 00:22:35.281 [opensc-pkcs11] reader-pcsc.c:1093:part10_get_vendor_product: id_vendor=25dd id_product=3111
P:6268; T:0x140653054314560 00:22:35.282 [opensc-pkcs11] reader-pcsc.c:1515:pcsc_detect_readers: returning with: 0 (Success)
```
OpenSC has: 
#define FEATURE_VERIFY_PIN_DIRECT        0x06
#define FEATURE_MODIFY_PIN_DIRECT        0x07
#define FEATURE_IFD_PIN_PROPERTIES       0x0A
#define FEATURE_GET_TLV_PROPERTIES       0x12
These were all found by the code at:
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L1170-L1208
But the   pcsc10_v2.02.08 says:
2.6.4 FEATURE_VERIFY_PIN_DIRECT / FEATURE_MODIFY_PIN_DIRECT
"1) class 2 readers without display for which no user feedback to the host is either required or supported."

It could be an OpenSC bug, that says it has a pin pad reader because it has FEATURE_VERIFY_PIN_DIRECT
but the wording is pcsc10_v2.02.08 implies DIRECT means if no display, don't (have/use) PIN PAD.  

OpenSC then goes on to handle one reader that has a problem:
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L1220-L1229

@frankmorgner any ideas on the PART10?

**Comment by frankmorgner** [comment]
I think the purpose of VERIFY_PIN_DIRECT is to simplify management of the pin entry on the reader. To my understanding VERIFY_PIN_START/FINISH, equally do the job, but additionally allow more control, e.g. with ABORT or GET_KEY_PRESSED.

If this is a CCID reader, @LudovicRousseau may want to disable propagation of the PIN pad support if it is broken...

**Comment by LudovicRousseau** [comment]
The Bit4id miniLector I have in my list is not, and does not declare, a pinpad feature
https://ccid.apdu.fr/ccid/shouldwork.html#0x25DD0x3111

But it looks like the firmware evolved.
@fedetft can you follow the procedure at https://ccid.apdu.fr/#CCID_compliant with your Bit4id miniLector please?

**Comment by fedetft** [comment]
Sure, here it is
[output.txt](https://github.com/OpenSC/OpenSC/files/6593720/output.txt)

**Comment by LudovicRousseau** [comment]
The reader now declares:
```
  bPINSupport: 0x03
   PIN Verification supported
   PIN Modification supported
```
If the reader really looks like the picture at https://www.bit4id.com/fr/lettore-di-smart-card-minilector-evo/ then I guess the reader does NOT have a pinpad :-)

I will try to contact bit4id.

**Comment by fedetft** [comment]
On the outside it looks just like the previous model, there is no pinpad.

**Comment by LudovicRousseau** [comment]
bit4id confirmed it is a firmware bug. The bug will be solved the the next firmware release.

You can try this CCID patch to also "fix" the problem.
https://salsa.debian.org/rousseau/CCID/-/commit/cf8c58e949ce70cb5cf67b743deb1fd9aaa66fc2

**Comment by fedetft** [comment]
I got the pin of my CNS, and confirm that with `enable_pinpad = false;` in `opensc.conf` I can use it to login with Firefox.
Thanks everyone for the support.

**Comment by frankmorgner** [comment]
glad to see at least some issue resolved. the rest of it (new versions of Italian CNS cards) should be fixed with #2371
